[masonry] Implement Item-Tolerance

stwrt opened the pull request:
#### 42f279a610d0655fd80032c2fa0968480ae96c0d
<pre>
[masonry] Implement Item-Tolerance
<a href="https://bugs.webkit.org/show_bug.cgi?id=301968">https://bugs.webkit.org/show_bug.cgi?id=301968</a>
<a href="https://rdar.apple.com/problem/164043151">rdar://problem/164043151</a>

Reviewed by Sammy Gill.

Implement item-tolerance for CSS Masonry.

<a href="https://drafts.csswg.org/css-grid-3/#placement-tolerance">https://drafts.csswg.org/css-grid-3/#placement-tolerance</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004.html
* Source/WebCore/Sources.txt
* Source/WebCore/WebCore.xcodeproj/project.pbxproj
* Source/WebCore/css/CSSProperties.json
* Source/WebCore/rendering/GridMasonryLayout.cpp
* Source/WebCore/rendering/style/RenderStyle.h
* Source/WebCore/rendering/style/RenderStyleInlines.h
* Source/WebCore/rendering/style/RenderStyleSetters.h
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h
* Source/WebCore/style/StyleBuilderCustom.h
* Source/WebCore/style/values/grid/StyleItemTolerance.cpp

Canonical link: <a href="https://commits.webkit.org/303096@main">https://commits.webkit.org/303096@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e847188b5a4a353706dde6f368d83f111798d08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131285 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138728 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83023 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b67bd2e3-83fa-44bd-97bd-21dfa0457f4b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3457 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100032 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67787 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d07a85f0-4155-43eb-ba75-11b4c399d3ea) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2613 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80733 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d81470c0-3e3e-42f8-b861-1501165f20c0) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2528 "Found 5 new test failures: imported/w3c/web-platform-tests/css/css-contain/content-visibility/touch-action-beside-display-locked-fixedpos-iframe-crash.html imported/w3c/web-platform-tests/html/interaction/focus/focus-input-type-switch.html imported/w3c/web-platform-tests/html/rendering/widgets/input-date-no-resize-on-hover.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-toggle-source.html imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/242 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81980 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111128 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35645 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141230 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3360 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108550 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108495 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27577 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2535 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113853 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56510 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3422 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32277 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3244 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66830 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3444 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3352 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/style/values/grid/StyleItemTolerance.h:
```diff
@@ -0,0 +1,85 @@
+    {
+    }
+
+    ItemTolerance(LengthPercentage<CSS::Nonnegative>&& lengthPercentage)
```
This should probably be CSS::NonnegativeUnzoomed

sammygill commented on Source/WebCore/style/values/grid/StyleItemTolerance.h:
```diff
@@ -0,0 +1,85 @@
+
+// <'item-tolerance'> = normal | <length-percentage [0,∞]> | infinite
+// https://drafts.csswg.org/css-grid-3/#item-tolerance
+struct ItemTolerance {
```
I *think* this should probably inherit from `LengthWrapperBase` in the same way `PreferredSize` does

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +194,86 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+            },
+            [](CSS::Keyword::Infinite) { }
+        );
+    }
```
Can this entire thing just be a `tolerance.switchOn()`?

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +194,86 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+            [](CSS::Keyword::Normal) { },
+            [&](const Style::LengthPercentage<CSS::Nonnegative>& lengthPercentage) {
+                lengthPercentage.switchOn(
+                    [&](const auto& length) {
```
I think if you make `ItemTolerance` inherit from `LengthWrapperBase` you can change this into a `switchOn` with the individual cases broken down like in `RenderBox::computeContentAndScrollbarLogicalHeightUsing`

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+
+    // For normal and length-percentage tolerances, find positions within tolerance of the shortest track
+    // Calculate tolerance value
+    auto referenceLength = gridAxisDirection() == Style::GridTrackSizingDirection::Columns
```
Not a big fan of the name `referenceLength`. Presumably this should at least be something like `contentBoxSize` or similar to be a bit more descriptive. I presume this is what percentages would resolve against but I don't actually see the spec call that out anywhere https://drafts.csswg.org/css-grid-3/#placement-tolerance

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+            return LayoutUnit { fixed.resolveZoom(m_renderGrid->style().usedZoomForLength()) };
+        },
+        [&](const typename Style::ItemTolerance::Percentage& percentage) -> LayoutUnit {
+            return LayoutUnit { percentage.value / 100.0f * referenceLength };
```
I think something like this should work
```suggestion
            return Style::evaluate<LayoutUnit>(percentage, referenceLength)
```

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+        : m_renderGrid->contentBoxLogicalWidth();
+
+    LayoutUnit toleranceValue = tolerance.switchOn(
+        [&](const CSS::Keyword::Normal&) -> LayoutUnit {
```
Since the lambda has `LayoutUnit` as the trailing return value maybe we can can then `return LayoutUnit { ... }` to `return { ....}`?

stwrt replied:
`computedFontSize()` returns a float, so we need to construct a `LayoutUnit`.

`return LayoutUnit { m_renderGrid->style().computedFontSize() };`

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+            return LayoutUnit { percentage.value / 100.0f * referenceLength };
+        },
+        [&](const typename Style::ItemTolerance::Calc& calc) -> LayoutUnit {
+            return LayoutUnit { calc.protectedCalculation()->evaluate(referenceLength, m_renderGrid->style().usedZoomForLength()) };
```
This could also probably be `Style::evaluate<LayoutUnit>` if you think it looks cleaner

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+        absoluteShortest = std::min(absoluteShortest, maxPosForTrack);
+    }
+
+    // Step 2: Find first position (in DOM order) within tolerance of shortest
```
I don't think I see the connection with the DOM order here.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+    // Step 1: Find the absolute shortest position across all tracks
+    auto maxStartingLine = gridAxisLines - itemSpanLength;
+    LayoutUnit absoluteShortest = LayoutUnit::max();
+    for (unsigned i = 0; i < maxStartingLine; i++) {
+        LayoutUnit maxPosForTrack;
+        for (unsigned lineOffset = 0; lineOffset < itemSpanLength; lineOffset++)
+            maxPosForTrack = std::max(maxPosForTrack, m_runningPositions[i + lineOffset]);
+        absoluteShortest = std::min(absoluteShortest, maxPosForTrack);
+    }
```
Since we're looking at just the different values of the running positions can this just be one iteration over the vector? i.e. why do we need the outer and inner loop?

stwrt replied:
The nested loops are there to handle items that span multiple tracks.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -194,18 +195,78 @@ void GridMasonryLayout::updateItemOffset(const RenderBox& gridItem, LayoutUnit o
+    unsigned smallestMaxPosLine = 0;
+    for (unsigned i = 0; i < maxStartingLine; i++) {
+        // Start from cursor position and wrap around to respect DOM order
+        auto startingLine = (m_autoFlowNextCursor + i) % maxStartingLine;
```
Is this modulo maxStartingLine or number of tracks?

stwrt replied:
This modulo is the max starting line to handle items that span multiple tracks.

sammygill reviewed (commented):
I see the tests has the rows as the stacking axis but could we also add some tests for columns as the stacking axis?

sammygill reviewed (approved):
Don't forget to fix up imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html!

stwrt committed:
[masonry] Implement Item-Tolerance
https://bugs.webkit.org/show_bug.cgi?id=301968
rdar://problem/164043151

Reviewed by Sammy Gill.

Implement item-tolerance for CSS Masonry.

https://drafts.csswg.org/css-grid-3/#placement-tolerance

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-001.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-002.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-003.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/item-tolerance-004.html
* Source/WebCore/Sources.txt
* Source/WebCore/WebCore.xcodeproj/project.pbxproj
* Source/WebCore/css/CSSProperties.json
* Source/WebCore/rendering/GridMasonryLayout.cpp
* Source/WebCore/rendering/style/RenderStyle.h
* Source/WebCore/rendering/style/RenderStyleInlines.h
* Source/WebCore/rendering/style/RenderStyleSetters.h
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h
* Source/WebCore/style/StyleBuilderCustom.h
* Source/WebCore/style/values/grid/StyleItemTolerance.cpp

Canonical link: https://commits.webkit.org/303096@main
